src/cio_file_unix.c

```diff
@@ -118,6 +118,14 @@ int cio_file_native_remap(struct cio_file *cf, size_t new_size)
 
 /* OSX mman does not implement mremap or MREMAP_MAYMOVE. */
 #ifndef MREMAP_MAYMOVE
+    /* OSX unmap/mmap pair should be succeeded for provided news_size
+     * to be grearter than cf->alloc_size case.
+     * If not, this sort of operation will be failed.
+     * Just return early here. */
+    if (new_size < cf->alloc_size) {
```

Comment:
if this means we cannot shrink a chunk "I think" we might have the following unhandled cases:

- A chunk is always created with extra size, when the Chunk is putted in a down state, the size needs to be adjusted to reduce the extra space added at the beginning.

please re-validate this

Reply:
Hmm..., got it. I'm re-evaluate this.

Reply:
As far as I observed, on macOS, I cannot run unmap/mmap without error less than page_size(16384).

Reply:
> * the size needs to be adjusted to reduce the extra space added at the beginning.

This could be caused memory corruption.

On the other hand, the following patch is also able to consider to handle mremap emulation on macOS:

```diff
diff --git a/src/cio_file.c b/src/cio_file.c
index 786a022..b4d5218 100644
--- a/src/cio_file.c
+++ b/src/cio_file.c
@@ -1130,6 +1130,17 @@ int cio_file_resize(struct cio_file *cf, size_t new_size)
     mapped_flag = cio_file_native_is_mapped(cf);
     mapped_size = cf->alloc_size;
 
+#ifdef __APPLE__
+    /* On macOS, when the size of new_size is less than pagesize,
+     * unmap/mmap operations will cause unexpected behavior such as
+     * memory corruptions.
+     * This could make diffucult to handle further operations.
+     * In this case, prohibit to shrink file size. */
+    if (cio_getpagesize() > new_size) {
+        return CIO_OK;
+    }
+#endif
+
 #ifdef _WIN32
     if (mapped_flag) {
         result = cio_file_native_unmap(cf);
```

However, this causes checksum corruptions.

Reply:
So, my thought is:

* Prevent to shrink less than page size (16384 bytes for macOS-arm64) on macOS.
* Call cio_file_native_resize due to accurate calculations for checksum

Reply:
maybe the re-map is not needed, because if it needs to be shrink the process can be:

- sync changes to disk
- unmap the memory region
- truncate the file to the desired size
- mmap() again if needed

Reply:
Yes. Intending to implement less than page size operation on macOS is dangerous.
For syncing contents of actual mmap-ed files, we just call msync to handle written characters in mmap-ed regions.

The root cause of this issue should be cause by less mmap-ed size than pagesize. Basically, the page size of macOS should be 4096 in Intel Mac, and should be 16384 on Apple Silicon. In macOS, there is no ability to shrink less than pagesize.
Other platforms can be used for mmap-ed files less than 4096/16384.

For Linux, we can use `mremap` to shrink mmap-ed region less than pagesize. This could be valid operation on Linux.
For Windows, we can use `CreateFileMappingA` and `MapViewOfFile` to map less than ordinary pagesize. This also could be valid operation on Windows.
Although, in macOS, there is no way to handle less than page size. 
It cause always 0xfffffffff(-1) on start to read mmap-ed contents....